ministryofjustice/correspondence_tool_staff
# CT-3172 complaints reporting

## Description
This PR implements open and closed cases reports for Offender SAR Complaints as a baseline. In view of the uncertainty around MI and to get the PR into review I've ignored the monthly report for now.

It also includes some small refactoring and minor changes, including adding separate predicate methods for identifying _any_ offender sar case, a standard one, or a complaint: 
* offender_sar?                       # could be either - use this when we don't care which
* offender_sar_standard?
* offender_sar_complaint?

We also address some issues found with case creation for the new correspondence type, namely separating out the session data for in-progress complaints vs standard offender sars, also fixing a bug I found with the Back link, so that it now correctly remains within offender SAR complaint instead of "jumping tracks" back to standard offender SAR. 

## Self-review checklist


### Screenshots

![image](https://user-images.githubusercontent.com/1161161/99511671-7a56e580-2980-11eb-80ee-573c301e1f4c.png)


### Related JIRA tickets
https://dsdmoj.atlassian.net/browse/CT-3172

### Deployment
Has both data and database migrations :) 

Run `rails db:migrate` and `rails data:migrate`

### Manual testing instructions
As Branston user: Viewing open cases list containing offender sar and offender sar complaint cases, "Download as CSV" should export the cases of both types according to the filters. (Filters bug on complaint fixed in Yikang's PR) 
Closed cases reports should contain only offender sar or offender sar complaints respectively
